src/components/multi_select_state_utils.js

```diff
@@ -40,12 +40,30 @@ export const getMinMaxIndexes = (currentIndex, firstItemShiftSelected) =>
 export const isWithin = (index, { minIndex, maxIndex }) =>
   index >= minIndex && index <= maxIndex;
 
-export const getNewSelectedItems = (itemId, items, selectedItems) => {
-  const sourceItems = items.filter(
-    item => item.id === itemId || findItem(item, selectedItems)
-  );
-  const destinationItems = selectedItems.filter(
+export const getNewSelectedItems = (
+  itemId,
+  items,
+  selectedItems,
+  keepSelectionOrder
+) => {
+  let alreadySelectedItems = [];
+  let sourceItems = [];
+  let destinationItems = [];
+
+  if (keepSelectionOrder) {
+    // In order to keep selection order on the list,
+    // First, iterate threw already selectedItems
+    alreadySelectedItems = selectedItems.filter(item => findItem(item, items));
+    sourceItems = items.filter(item => item.id === itemId);
+  } else {
+    sourceItems = items.filter(
+      item => item.id === itemId || findItem(item, selectedItems)
+    );
+  }
+
+  destinationItems = selectedItems.filter(
```

Comment:
Did you find case when destinationItems is not empty? becouse I can't find use case for it, I think it's redundancy

Reply:
No, I haven't. Since it was some legacy code, I was frightened to remove it.